shared/clock/clock.cpp

```diff
@@ -2,31 +2,39 @@
 #include "clock.hpp"
 
 #include <chrono>
-#include <mutex>
 
 namespace utility::clock {
 
-    // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
-    double custom_rtf = 1.0;  // real time factor
     // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
     NUClear::base_clock::time_point last_update = NUClear::base_clock::now();
-    NUClear::base_clock::time_point state       = last_update;
-    std::mutex mutex;
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+    NUClear::base_clock::time_point epoch = last_update;
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+    double rtf = 1.0;  // real time factor
+
+    void update_rtf(const double& new_rtf) {
+        auto now = NUClear::base_clock::now();
+
+        // Since we are updating our rtf we need to advance our state so the new deltas will be calculated properly
+        // We set the variables in this specific order to mimise the error that will occur if the threading reads as we
+        // are writing By reducing the delta between state and now any changes in rtf will have a minimal change on
+        // delta. The old rtf is used since it is assume that since the last update the old rtf was in effect and
+        // only now with future calculations the new rtf will be used
+        epoch = epoch
+                + std::chrono::duration_cast<std::chrono::steady_clock::duration>(
+                    (now - last_update) * rtf);  // set before we update the variables
+        last_update = now;
+        rtf         = new_rtf;
```

Comment:
The more I think about it, the more I realise this isn't the way to go about this, you're going to get strange results as you go.
I think you should store the epoch, last_update and rtf in a struct (local to this c++ file) and have an array of two of them, then swap which one is "live" and which one is not using atomics, that way you'll have an invalid time much less often

Reply:
I think I get what you mean, I'll see about implementing this. It seems to work ok enough in practice but I can see how it might have problems still.